packages/polywrap-client-config-builder/polywrap_client_config_builder/client_config_builder.py

```diff
@@ -0,0 +1,149 @@
+        """
+        return self.config
+
+    def get_envs(self)-> Dict[Uri, Dict[str, Any]]:
+        return self.config.envs
+        
+    def set_env(self, env: Env={'test':'tested'}, uri=Uri("wrap://ens/eth.plugin.one"))-> IClientConfigBuilder:
+        # TODO:  This is a temporary solution to the problem of the Uri / Env not being able to pass as None initially 
+        # should be (self, env: Env=None, uri: Uri=None):  but that causes an error
+        if (env or uri) is None:
+            raise KeyError("Must provide both an env or uri")
+        self.config.envs[uri] = env
+        return self
+
+        # for key in self.config.envs[uri].keys():
+        #     if key in env.keys():
+        #         self.config.envs[uri][key] = env[key]
+        #     else:
+        #         self.config.envs[uri][key] = None
+        # return self
+
+    def add_env(self, env: Env = None , uri: Uri = None)-> IClientConfigBuilder:
+        if (env or uri) is None:
+            raise KeyError("Must provide both an env or uri")
+        self.config.envs[uri] = env
+        return self
+
+    def add_envs(self, envs: List[Env]) -> IClientConfigBuilder:
+        """
+        Adds a list of environments (each in the form of an `Env`) for a given uri
+        """
+        for env in envs:
+            self.add_env(env)
+        return self
+
+    def add_interface_implementations(self, interface_uri: Uri, implementations_uris: List[Uri]) -> IClientConfigBuilder:
+        """
+        Adds a list of implementations (each in the form of an `Uri`) for a given interface
+        """
+        if interface_uri is None:
+            raise ValueError()
+        if interface_uri in self.config.interfaces.keys():
+            self.config.interfaces[interface_uri] = self.config.interfaces[interface_uri] + implementations_uris
+        else:
+            self.config.interfaces[interface_uri] = implementations_uris
+        return self
+
+    def add_wrapper(self, wrapper_uri: Uri) -> IClientConfigBuilder:
+        """
+        Adds a wrapper to the list of wrappers
+        """
+        self.config.wrappers.append(wrapper_uri)
+        return self
+
+    def add_wrappers(self, wrappers_uris: List[Uri]) -> IClientConfigBuilder:
+        """
+        Adds a list of wrappers to the list of wrappers
+        """
+        for wrapper_uri in wrappers_uris:
+            self.add_wrapper(wrapper_uri)
+        return self
+
+    def remove_wrapper(self, wrapper_uri: Uri) -> IClientConfigBuilder:
+        """
+        Removes a wrapper from the list of wrappers
+        """
+        self.config.wrappers.remove(wrapper_uri)
+        return self
+
+    def set_resolver(self, uri_resolver) -> IClientConfigBuilder:
+        """
+        Sets a single resolver for the `ClientConfig` object before it is built
+        """
+        self.config.resolver = uri_resolver
+        return self
+
+class ClientConfigBuilder(BaseClientConfigBuilder):
```

Comment:
Should this be removed?

Reply:
The reason this was left in was that initially the ClientConfigBuilder is the object we call on the JS implementation

When we have defaults configs we can add them to this object [for example](https://github.com/polywrap/toolchain/blob/origin-dev/packages/js/client-config-builder/src/ClientConfigBuilder.ts)

So the tests we instantiate 
`ccb = ClientConfigBuilder`, 

and not 
`bccb = BaseClientConfigBuilder`. 

I understand the BaseCCB is the benchmark functionality a client should have, while ClientConfigBuilder would be particular implementations of that BaseCCB

this also brings up the question on yesterday's call of how to structure the different client configurations.

Let me know if we should remove or rename them it and i'll update the classes and the tests to fit the correct implementation

Reply:
Gotcha, so what I would suggest would be the following:
- In the `BaseClientConfigBuilder` let's implement the main methods (basically, all, except for `build` - this will be the only method without implementation)
- In the `ClientConfigBuilder` we only implement `build`, inspired on [this method](https://github.com/polywrap/toolchain/blob/origin-dev/packages/js/client-config-builder/src/ClientConfigBuilder.ts#L26-L46) from JS client. Note that currently the resolvers are not implemented so you can just return a mock object, let's just focus on the architecture now

Reply:
Just for context, currently we have `IClientConfigBuilder` which is the abstract class, `BaseClientConfigBuilder` which is the one that implements and `ClientConfigBuilder` is just there to have the same API. With the suggested changes this will make that `IClientConfigBuilder` doesn't exists anymore and the `BaseClientConfigBuilder` will inherit from `ABC` (meaning that is the abstract class), here, we will implement all methods except build (making them abstract, in case the use that inherit from this class can define custom behavior); and `ClientConfigBuilder` will only implement `build`